backends/dpdk/dpdkArch.h

```diff
@@ -443,6 +443,42 @@ class ConvertInternetChecksum : public PassManager {
     }
 };
 
+/* This pass collects PSA register declaration instances and push them to a map
+ * for emitting to the .spec file later */
+class CollectRegisterDeclaration : public Inspector {
+    std::map<const IR::Declaration_Instance *, cstring> *reg_map;
+    P4::TypeMap *typeMap;
+
+  public:
+    CollectRegisterDeclaration(
+        std::map<const IR::Declaration_Instance *, cstring> *reg_map, P4::TypeMap *typeMap)
+        : reg_map(reg_map) , typeMap(typeMap) {}
+
+    bool preorder(const IR::Declaration_Instance *d) override {
+        if (d->type->is<IR::Type_Specialized>()) {
+            auto type = d->type->to<IR::Type_Specialized>();
+            auto externTypeName = type->baseType->path->name.name;
+            if (externTypeName == "Register"){
+                if (findContext<IR::P4Control>()){
+                    if (d->arguments->size() != 1 and d->arguments->size() != 2 ) {
+                        ::error("%1%: expected size and optionally init_val as arguments", d);
+                     }
+                    reg_map->emplace(d, d->name);
+                }
```

Comment:
Han, referring to PSA specification (Table 5, Section 7.1) , Register extern may only be instantiated and called from Ingress and Egress. Should register instantiation be allowed outside of Ingress and Egress?

Reply:
Yes, you are right. Sorry about the confusion. A warning as Mihai suggested would be great.

Reply:
Just to be clear, the PSA spec says that implementations are allowed to reject programs that instantiate such externs in other places.  It does not REQUIRE that an implementation reject such instantiations.  A PSA implementation is allowed to support such externs in other places, too.  There are explicit statements saying so in Section 7.1 of the PSA spec.

Reply:
Thanks for clarifying this Andy. I think we can restrict the use of externs in control blocks for now (For all the good reasons  mentioned in the PSA spec such as portability, performance and use in multi-pipeline device). If there is a need to allow it in future, we can add that support.

Reply:
It seems reasonable to restrict the generality of such things in an initial implementation, if you wish.  If you are concerned about portability from programs written for DPDK to other PSA targets, I wouldn't worry overly much about that -- you can simply say in the documentation for a DPDK implementation that supports accessing the same registers in both ingress and egress that this might be uncommon for other PSA implementations, and not to rely upon that capability existing there if they care about portability.  But there doesn't seem to be a strong reason to prevent people with DPDK as a primary target from being more general if they wish, and you are willing to support it.

Reply:
Agree. I updated the implementation to remove this restriction.